File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int n = s.size();
        int left = 0;
        int right  = 0;
        int count =0;
        unordered_map<char,int>st;
        while(right <n){
            if(st.size()<3){
                if(s[right]=='a'||s[right]=='b'||s[right]=='c'){
                    st[s[right]]++;
                }
                right++;
            }
            else{
                count+= (n-right+1);
                while(left<right&&st.size()>=3){
                    if(st.find(s[left])!=st.end()){
                        st[s[left]]--;
                        if(st[s[left]]==0) st.erase(s[left]);
                    }
                    if(st.size()==3) count+= (n-right+1);
                    left++;
                }
            }
        }
        while(left<right&&st.size()>=3){
            if(st.size()==3) count++;
            if(st.find(s[left])!=st.end()){
                st[s[left]]--;
                if(st[s[left]]==0) st.erase(s[left]);
            }
            left++;
        }
        return count;

    }
};

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
from collections import defaultdict
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        left = 0
        right = 0
        a =0
        b= 0
        c = 0
        cnt =0
        while right<n:
            if s[right]=='a':
                a+=1
            if s[right]=='b':
                b+=1
            if s[right]=='c':
                c+=1
            if a>0 and b>0 and c>0:
                cnt += (n - right)
                while left<right and a>0 and b>0 and c>0:
                    if s[left]=='a':
                        a-=1
                    if s[left]=='b':
                        b-=1
                    if s[left]=='c':
                        c-=1
                    if a>0 and b>0 and c>0:
                        cnt += (n - right)
                    left+=1
            right+=1
        if a>0 and b>0 and c>0:
            cnt += (n - right)
        return cnt